---
title: "Exploring TCGA"
output: html_notebook
---

## Load dependencies
```{r}
library(tidyverse)
```

# 1 Reading & Cleaning

## Define where the data lives
```{r}
## Define data paths
in_path_varscan = 'data/maf/6c93f518-1956-4435-9806-37185266d248/TCGA.BRCA.varscan.6c93f518-1956-4435-9806-37185266d248.DR-10.0.somatic.maf.gz'
in_path_muse = 'data/maf/b8ca5856-9819-459c-87c5-94e91aca4032/TCGA.BRCA.muse.b8ca5856-9819-459c-87c5-94e91aca4032.DR-10.0.somatic.maf.gz'
in_path_ss = 'data/maf/7dd592e3-5950-4438-96d5-3c718aca3f13/TCGA.BRCA.somaticsniper.7dd592e3-5950-4438-96d5-3c718aca3f13.DR-10.0.somatic.maf.gz'
in_path_mutect = 'data/maf/995c0111-d90b-4140-bee7-3845436c3b42/TCGA.BRCA.mutect.995c0111-d90b-4140-bee7-3845436c3b42.DR-10.0.somatic.maf.gz'
```

## Reading data
```{r}
df = read_tsv(
  in_path_mutect,
  comment = '#'
)
```


## Preview data
```{r}
print(dim(df))
head(df, 10)
```


## Read & concatenate multiple files
```{r}
## Define column types
col_types = cols_only(
  Chromosome = 'c',
  Start_Position = 'i',
  End_Position = 'i',
  SYMBOL = 'c',
  Reference_Allele = 'c',
  Allele = 'c', 
  Variant_Classification = 'c', 
  IMPACT = 'c',
  Variant_Type = 'c',
  Tumor_Sample_Barcode = 'c'
)

## Concatenate vertically
in_paths = c(
  in_path_varscan,
  in_path_muse,
  in_path_ss,
  in_path_mutect
)
df = bind_rows(lapply(
  in_paths, 
  read_tsv, 
  col_types = col_types,
  comment = '#'
))

print(dim(df))
head(df)
```

## Rename and reorder columns
```{r}
## Rename
col_name_map = list(
  CHR = 'Chromosome',
  START = 'Start_Position',
  END = 'End_Position',
  GENE = 'SYMBOL',
  REF = 'Reference_Allele',
  ALT = 'Allele',
  CLASS = 'Variant_Classification',
  IMPACT = 'IMPACT',
  TYPE = 'Variant_Type',
  BARCODE = 'Tumor_Sample_Barcode'
)
df = rename(df, !!! col_name_map)

## Reorder
keep_cols = names(col_name_map)
df = select(df, keep_cols)

head(df)
```

## Fill missing data
```{r}
df$GENE[is.na(df$GENE)] = 'INTERGENIC'
```

## Create new columns from string ops
```{r}
df$SAMPLE = str_sub(df$BARCODE, 1, 12)
df$MUTATION = paste(
  df$CHR, 
  df$GENE,
  df$START,
  df$END,
  df$REF,
  df$ALT,
  sep = ':'
)
```

## Remove duplicates
```{r}

## Remove duplicate rows
df = unique(df)

## Remove duplicate based on column
df_mut = df[!duplicated(df$MUTATION),]

print(dim(df))
print(dim(df_mut))
```

## Remove columns
```{r}
drop_cols = c('BARCODE', 'SAMPLE')
df_mut = select(df_mut, -one_of(drop_cols))

print(dim(df_mut))
head(df_mut)
```

## Writing files
```{r}
write_delim(
  df_mut, 
  'data/out/df_r.tsv', 
  delim = '\t'
)
```


## Read & write binary files
```{r}

rds_path = 'data/out/df.rds'

## Write
saveRDS(df_mut, rds_path)

## Read
df = readRDS(rds_path)
```


## Long to wide
```{r}

```

## Add info from 1 df to another (merge)
```{r}

```

## Wide to long
```{r}

```

## Free memory
```{r}

```

# Counting 

## Count unique elements
```{r}

```

## Count factor levels
```{r}

```

## Summarize by group
```{r}
df_summary = df %>%
  group_by(GENE) %>%
  summarise(
    N_MUTATIONS = n(),
    N_UNIQUE_MUTATIONS = length(unique(MUTATION)),
    SAMPLES = length(unique(BARCODE))
)
```

## Sorting
```{r}
df_summary = df_summary %>%
  arrange(desc(SAMPLES), N_MUTATIONS)
```


## Summarize by multiple groups
```{r}
df_impact = df %>%
  group_by(GENE, IMPACT) %>%
  summarise(COUNT = length(unique(MUTATION)))
```

# Visualize

```{r}
img_dir = 'data/out/img/'
```


## Ranking
```{r}

```

## Bar plot
```{r}

```

## Subplots
```{r}

```

## Count vs identity
```{r}

```


## Plot with grouping
```{r}
## Filter to plotting data
genes = df_summary$GENE[1:10]
df_plt = df_impact %>%
  filter(GENE %in% genes)

## Reorder for plot
df_plt$GENE = factor(df_plt$GENE, levels = genes)

p = ggplot(data = df_plt) +
  geom_bar(
    aes(x=GENE, y=COUNT, fill=IMPACT),
    stat = 'identity',
    position = 'dodge'
  ) +
  scale_fill_brewer(
    type='qual',
    palette = 'Set1'
  )
p
```

## Faceted plot
```{r}

```

## Saving
```{r}

```
